pkg/kademlia/dht.go

```diff
@@ -5,37 +5,178 @@ package kademlia
 
 import (
 	"context"
+	"crypto/rand"
+	"fmt"
+	"strings"
 
-	"storj.io/storj/protos/overlay"
+	bkad "github.com/coyle/kademlia"
+	"github.com/zeebo/errs"
+
+	proto "storj.io/storj/protos/overlay"
 )
 
+// NodeErr is the class for all errros petaining to node operations
+var NodeErr = errs.Class("node error")
+
+//TODO: shouldn't default to TCP but not sure what to do yet
+var defaultTransport = proto.NodeTransport_TCP
+
 // Kademlia is an implementation of kademlia adhering to the DHT interface.
 type Kademlia struct {
+	rt             RoutingTable
+	bootstrapNodes []proto.Node
+	ip             string
+	port           string
+	stun           bool
+	dht            *bkad.DHT
+}
+
+// NewKademlia returns a newly configured Kademlia instance
+func NewKademlia(bootstrapNodes []proto.Node, ip string, port string, stun bool) Kademlia {
+	bb := convertProtoNodes(bootstrapNodes)
+	bdht, _ := bkad.NewDHT(&bkad.MemoryStore{}, &bkad.Options{
+		ID:             []byte(newID()),
+		IP:             ip,
+		Port:           port,
+		BootstrapNodes: bb,
+	})
+
+	rt := RouteTable{
+		ht:  bdht.HT,
+		dht: bdht,
+	}
+
+	return Kademlia{
+		rt:             rt,
+		bootstrapNodes: bootstrapNodes,
+		ip:             ip,
+		port:           port,
+		stun:           stun,
+		dht:            bdht,
+	}
 }
 
-// GetNodes returns all nodes from a starting node up to a maximum limit
-func (k Kademlia) GetNodes(ctx context.Context, start string, limit int) ([]overlay.Node, error) {
-	return []overlay.Node{}, nil
+// GetNodes returns all nodes from a starting node up to a maximum limit stored in the local routing table
+func (k Kademlia) GetNodes(ctx context.Context, start string, limit int) ([]proto.Node, error) {
```

Comment:
can the `start` id be bytes instead of a string? seems weird to use a string but maybe it's okay

Reply:
so actually, this is an interesting comment. strings are expected to be printable and []bytes are not necessarily, yeah, but there's another useful difference, which is that strings are immutable and []bytes are not. you can use a string as a key in a map (not a byte slice), and you can guarantee that no one will change the bytes within a string out from under you. for this reason, i often actually do use strings for byte data and use the two interchangably. for ids and so on i almost always use strings, even if the data isn't hex/base64 and is straight binary data.

that said, i get that makes it weird from a human perspective if you're expecting printable characters and get garbage from a string, so in a previous codebase I made a package called types where we had a `types.Binary` type, which was really just a `string` under the covers, but that way other programmers knew i meant for it to be an immutable set of bytes. might be worth considering here?

Reply:
an unrelated comment - as soon as we try and change stuff on the `k Kademlia` receiver struct in the future, we're going to kick ourselves for hours trying to hunt down the bug of why the fields aren't changing until we realize it should be a `k *Kademlia` pointer receiver. i used to do receiver types by copy sometimes until i got sick of inevitably trying to figure out that bug, so now i think by reference receivers almost always make the most sense. i think it makes sense here